src/DefaultTokenUriResolver.sol

```diff
@@ -0,0 +1,600 @@
+import {IJBPaymentTerminal} from "@jbx-protocol/juice-contracts-v3/contracts/interfaces/IJBPaymentTerminal.sol";
+import {JBTokens} from "@jbx-protocol/juice-contracts-v3/contracts/libraries/JBTokens.sol";
+import {JBCurrencies} from "@jbx-protocol/juice-contracts-v3/contracts/libraries/JBCurrencies.sol";
+import {IJBController, IJBDirectory, IJBFundingCycleStore} from "@jbx-protocol/juice-contracts-v3/contracts/interfaces/IJBController.sol";
+import {IJBOperatorStore} from "@jbx-protocol/juice-contracts-v3/contracts/interfaces/IJBOperatorStore.sol";
+import {IJBPayoutRedemptionPaymentTerminal} from "@jbx-protocol/juice-contracts-v3/contracts/interfaces/IJBPayoutRedemptionPaymentTerminal.sol";
+import {IJBSingleTokenPaymentTerminalStore, IJBSingleTokenPaymentTerminal} from "@jbx-protocol/juice-contracts-v3/contracts/interfaces/IJBSingleTokenPaymentTerminalStore.sol";
+import {JBPayoutRedemptionPaymentTerminal} from "@jbx-protocol/juice-contracts-v3/contracts/abstract/JBPayoutRedemptionPaymentTerminal.sol";
+import {IJBProjects} from "@jbx-protocol/juice-contracts-v3/contracts/interfaces/IJBProjects.sol";
+import {IJBProjectHandles} from "@jbx-protocol/project-handles/contracts/interfaces/IJBProjectHandles.sol"; // Needs updating when NPM is renamed to /juice-project-handles
+import {JBOperatable} from "@jbx-protocol/juice-contracts-v3/contracts/abstract/JBOperatable.sol";
+import {JBUriOperations} from "./Libraries/JBUriOperations.sol";
+import {Theme} from "./Structs/Theme.sol";
+import {Base64} from "base64-sol/base64.sol";
+import {Strings} from "@openzeppelin/contracts/utils/Strings.sol";
+import {Font, ITypeface} from "typeface/interfaces/ITypeface.sol";
+
+// // ENS RESOLUTION (currently unused)
+// interface IReverseRegistrar {
+//     function node(address) external view returns (bytes32);
+// }
+
+// interface IResolver {
+//     function name(bytes32) external view returns (string memory);
+// }
+
+contract StringSlicer {
+    // This function is in a separate contract so that TokenUriResolver can pass it a string memory and we can still use Array Slices (which only work on calldata)
+    function slice(
+        string calldata _str,
+        uint256 _start,
+        uint256 _end
+    ) external pure returns (string memory) {
+        return string(bytes(_str)[_start:_end]);
+    }
+}
+
+contract DefaultTokenUriResolver is IJBTokenUriResolver, JBOperatable {
+    using Strings for uint256;
+    StringSlicer slice = new StringSlicer();
+
+    event Log(string message);
+    event ThemeSet(uint256 projectId, Theme theme);
+    error InvalidTheme();
+
+    IJBFundingCycleStore public fundingCycleStore;
+    IJBProjects public projects;
+    IJBDirectory public directory;
+    IJBTokenStore public tokenStore;
+    IJBSingleTokenPaymentTerminalStore public singleTokenPaymentTerminalStore;
+    IJBController public controller;
+    IJBProjectHandles public projectHandles;
+    ITypeface public capsulesTypeface; // Capsules typeface
```

Comment:
I think these can all be `immutable`

Reply:
Might as well because there's no setters to mutate them. Overall I'm not confident in how I'm getting Terminal addresses etc in the constructor and getUri. I think recent controller migration breaks the parts of the Default Resovler

Reply:
Fixed [c1f9f8c](https://github.com/jbx-protocol/juice-token-resolver/pull/1/commits/c1f9f8c28c09c01c3efc3c14add63cefe35171f1)